app.js

```diff
@@ -1,54 +1,194 @@
+    // If there are no vegetables that match, just return an empty array.
+    res.status(200).json(filteredVeg);
+  }
+});
+
+app.get('/vegetables/:id', (req, res, next) => {
+  const { vegetables } = data;
+  const { id } = req.params;
+   //find the first result in array where vegetable ID === req.params.id; there should only ever be one match since IDs are unique
+  const vegetable = vegetables.find(veggie => veggie.id === id);
+  // if a match is not found
+  if (!vegetable) {
+    // return a 404 status and error message of ID not found
+    const message = `Could not find vegetable with ID of ${id}`;
+    next({ status: 404, message });
+  }
+  // otherwise, as match is found
+  // return status 200, vegetable detail in response json
+  res.status(200).json(vegetable);
+});
+
+app.post('/vegetables', helpers.validate, (req, res, next) => {
+  const { vegetables } = data;
+  // create object vegetable with id, spread of body (name, price). id is generated by shortid
+  const vegetable = { id: generateId(), ...req.body };
+  // add vegetable object to vegetables array
+  vegetables.push(vegetable);
+  // return status 201, vegetable detail in response json
+  res.status(201).json(vegetable);
+});
+
+app.delete('/vegetables/:id', (req, res, next) => {
+  const { vegetables } = data;
+  const { id } = req.params;
+  //find the first result in array where vegetable ID === req.params.id; there should only ever be one match since IDs are unique
+  const vegetable = vegetables.find(veg => veg.id === id);
+  // if a match is not found
+  if (!vegetable) {
+    // return a 404 status and error message of ID not found
+    const message = `Could not find vegetable with ID of ${id}`;
+    next({status: 404, message});
+  } else {
+    // otherwise, as match is found
+    // find the index of vegetable where id matches req.params.id (again, only one match) and splice this index from vegetables array.
+    // then return updated array with 200 response code
+    vegetables.splice(vegetables.findIndex(veg => veg.id === id),1);
+    res.status(200).json(vegetable);
+  }
+});
+
+app.put('/vegetables/:id', (req, res, next) => {
+  const { vegetables } = data;
+  const { id } = req.params;
+  //find the first result in array where vegetable ID === req.params.id; there should only ever be one match since IDs are unique
+  const vegetable = vegetables.find(veg => veg.id === id);
+   // if a match is not found
+   if (!vegetable) {
+    // return a 404 status and error message of ID not found
+    const message = `Could not find vegetable with ID of ${id}`;
+    next({status: 404, message})
+  } else {
+    // otherwise, as match is found
+    // find the index of vegetable where id matches req.params.id (again, only one match) and update name and price values using req.params.
+    // then return updated vegetable with 201 response code
+    let indexOfVeg = vegetables.findIndex(veg => veg.id === id);
+    vegetables[indexOfVeg] = {id: id, ...req.body}

```

Comment:
This totally works, but the instructions technically say to throw an error if all of the keys are not present.